compiler/noirc_evaluator/src/ssa/opt/array_get.rs

```diff
@@ -0,0 +1,156 @@
+use crate::ssa::{
+    ir::{
+        dfg::CallStack, function::Function, instruction::Instruction, map::Id, types::Type,
+        value::ValueId,
+    },
+    ssa_gen::Ssa,
+};
+use fxhash::FxHashMap as HashMap;
+
+impl Ssa {
+    // Given an original IfElse instruction is this:
+    //
+    //     v10 = if v0 then v2 else if v1 then v3
+    //
+    // and a later ArrayGet instruction is this:
+    //
+    //     v11 = array_get v4, index v4
+    //
+    // we optimize it to this:
+    //
+    //     v12 = array_get v2, index v4
+    //     v13 = array_get v3, index v4
+    //     v14 = if v0 then v12 else if v1 then v13
+    #[tracing::instrument(level = "trace", skip(self))]
+    pub(crate) fn array_get_optimization(mut self) -> Self {
+        for function in self.functions.values_mut() {
+            // This should match the check in flatten_cfg
+            if let crate::ssa::ir::function::RuntimeType::Brillig = function.runtime() {
+                continue;
+            }
+
+            Context::default().optimize_array_get(function);
+        }
+
+        self
+    }
+}
+
+#[derive(Default)]
+struct Context {
+    // Given an IfElse instruction, here we map its result to that instruction.
+    // We only capture such values if the IfElse values are arrays.
+    result_if_else: HashMap<ValueId, Id<Instruction>>,
+}
+
+impl Context {
+    fn optimize_array_get(&mut self, function: &mut Function) {
+        let block = function.entry_block();
+        let dfg = &mut function.dfg;
+        let instructions = dfg[block].take_instructions();
+
+        for instruction_id in instructions {
+            let instruction = dfg[instruction_id].clone();
```

Comment:
If I do this instead:

```rust
let instruction = &dfg[instruction_id];
```

then it doesn't let me call `dfg.insert_instruction_and_results` later on because that's mutable and here we are taking an immutable reference to `dfg`.

What's the best way to improve this? (I tried to take a look at other passes but they were much more complex than this pass and I couldn't find a loop similar to this one where new instructions are added in the middle of the loop)

Reply:
Oh, I just found another pass that loops through instructions and clones them:

https://github.com/noir-lang/noir/blob/4e2c08d545178a3b495e0acaaad9f8c51f5d3418/compiler/noirc_evaluator/src/ssa/opt/constant_folding.rs#L175

So maybe cloning here isn't that bad after all?

Reply:
Cloning instructions isn't too bad. Cost-wise some of them may clone a `Vec<ValueId>` but that's about it. We may want to try to remove these for the future when optimizing the compiler but for now we're preferring readability & correctness.